llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -2742,7 +2742,12 @@ static Instruction *foldSelectWithFCmpToFabs(SelectInst &SI,
     // Note: We require "nnan" for this fold because fcmp ignores the signbit
     //       of NAN, but IEEE-754 specifies the signbit of NAN values with
     //       fneg/fabs operations.
-    if (!SI.hasNoSignedZeros() || !SI.hasNoNaNs())
+    if (!SI.hasNoNaNs())
+      return nullptr;
+
+    bool FunctionHasNoSignedZeroes =
+        SI.getParent()->getParent()->hasFnAttribute("no-signed-zeros-fp-math");
```

Comment:
We shouldn't spread additional uses of the attribute. This should only use the flags

Reply:
I'm unable to get it to work with floating point flags. `instcombine` requires `select` to have `nsz` flag to be able to fold a combination of `fcmp` `fneg` and `select` into a `fabs` intrinsic. But with current semantics, we can't safely propagate `nsz` to `select`. I might not be able to fully explain the problem in single comment but everything is well documented at #51601 particularly [this comment](https://github.com/llvm/llvm-project/issues/51601#issuecomment-981047527).

Reply:
In principle, I agree with @arsenm about not introducing additional uses of the attribute. I said in #51601 that I couldn't see a way around it, but @nikic's suggestion about setting the flag somewhere based on the attribute might be a way around this. I can't see a reason that it wouldn't be legal for SROA to set the nsz flag when it eliminates the stores and loads in this example: https://godbolt.org/z/fahb8oM9o

Reply:
Unfortunately, this is also the wrong way to test for the attribute. For some reason, these are specified as string booleans. You should use getValueAsBool.

Reply:
I tried 
```
bool FnHasNszAttr = SI.getFunction()->getFnAttribute("no-signed-zeros-fp-math").getValueAsBool();
```
But this is returning false while below code is returning true
```
bool FnHasNszAttr = SI.getFunction()->hasFnAttribute("no-signed-zeros-fp-math")
```
Is there anything I am misinterpreting?